impeller/entity/contents/atlas_contents.cc

```diff
@@ -81,18 +85,133 @@ const SamplerDescriptor& AtlasContents::GetSamplerDescriptor() const {
 bool AtlasContents::Render(const ContentContext& renderer,
                            const Entity& entity,
                            RenderPass& pass) const {
-  if (texture_ == nullptr) {
+  if (texture_ == nullptr || blend_mode_ == BlendMode::kClear) {
     return true;
   }
 
-  using VS = AtlasFillVertexShader;
-  using FS = AtlasFillFragmentShader;
+  if (blend_mode_ == BlendMode::kSource || colors_.size() == 0) {
+    return RenderTexture(renderer, entity, pass, alpha_);
+  }
+  if (blend_mode_ == BlendMode::kDestination) {
+    return RenderColors(renderer, entity, pass, alpha_);
+  }
 
-  const auto texture_size = texture_->GetSize();
-  if (texture_size.IsEmpty()) {
+  // Ensure that we use the actual computed bounds and not a cull-rect
+  // approximation of them.
+  auto coverage = ComputeBoundingBox();
+  auto size = coverage.size;
+
+  // Simple blends.
+  if (blend_mode_ < BlendMode::kScreen) {
+    auto subpass_texture = renderer.MakeSubpass(
+        ISize::Ceil(size),
+        [&contents = *this, &coverage](const ContentContext& renderer,
+                                       RenderPass& pass) -> bool {
+          Entity sub_entity;
+          sub_entity.SetBlendMode(BlendMode::kSourceOver);
+          sub_entity.SetTransformation(
+              Matrix::MakeTranslation(Vector3(-coverage.origin)));
+          if (!contents.RenderColors(renderer, sub_entity, pass, 1.0)) {
+            return false;
+          }
+          return contents.RenderTexture(renderer, sub_entity, pass, 1.0, true);
+        });
+    auto contents = ColorFilterContents::MakeBlend(
+        blend_mode_, {FilterInput::Make(subpass_texture)});
+    contents->SetAlpha(alpha_);
+    return contents->Render(renderer, entity, pass);
+  }
+
+  auto dst_texture = renderer.MakeSubpass(
+      ISize::Ceil(size),
+      [&contents = *this, &coverage](const ContentContext& renderer,
+                                     RenderPass& pass) -> bool {
+        Entity sub_entity;
+        sub_entity.SetBlendMode(BlendMode::kSourceOver);
+        sub_entity.SetTransformation(
+            Matrix::MakeTranslation(Vector3(-coverage.origin)));
+        return contents.RenderColors(renderer, sub_entity, pass, 1.0);
+      });
+  auto src_texture = renderer.MakeSubpass(
+      ISize::Ceil(size),
+      [&contents = *this, &coverage](const ContentContext& renderer,
+                                     RenderPass& pass) -> bool {
+        Entity sub_entity;
+        sub_entity.SetBlendMode(BlendMode::kSourceOver);
+        sub_entity.SetTransformation(
+            Matrix::MakeTranslation(Vector3(-coverage.origin)));
+        return contents.RenderTexture(renderer, sub_entity, pass, 1.0);
+      });
```

Comment:
Rather than rendering the textures here, any Contents can be passed to a filter as a ContentsFilterInput. Doing so has some nice performance benefits -- for example, these inputs evaluate lazily, and filters won't fetch the snapshot unless it's actually needed.
Also, the `ContentsFilterInput` evaluation uses `Contents::RenderToSnapshot`, which allows contents to make optimizations that avoid  creating a new texture in some cases by deferring parts of the rendering down the filter chain via other snapshot fields, like the transform.

Reply:
Makes sense, I should probably create a subcontents for both the src and dst and then I should be able to combine all of these cases into one. Perhaps that will fix the issue with destination alpha looking off, but any ideas on that?

Reply:
Though for vertices, I'll still probably need to manually create the texture in order to implement texture coordinate support.

Reply:
ahh yeah, re-arranging it this way looks much better. Alpha on dest looks like skia now